ONSdigital/eq-publisher-v3
# Piping tests

> BEFORE MAKING YOUR PR... Check the test coverage!

### What is the context of this PR?

Updating the Piping & Transformations to align Author with the new V3 schema. 
Rough overview of the changes can be found here:
https://github.com/ONSdigital/eq-schema-validator/blob/v3/doc/decisions/0001-define-piping-in-a-structured-way.md

This ticket is not to implement all the new ways to transform things but to align our current transformations i.e. dates, currency, numbers and text.

## How to test
**Requires local `eq-questionnaire-runner` (`eq-q-r`)** 
**branch: `fix_for_monthyear_piping`**

**Requires local `eq-questionnaire-validator` (`eq-q-v`)** 
**branch: `update_for_month_year_type`**

**Run`eq-q-v` with AJV**
- In `eq-q-v` root run: 
    - `cd ajv`
    - `DEBUG=validator node app.js`
- `eq-q-v` should be running on PORT 5001

- In `publisher-v3` add a `.env` file at project root if not already there
- Add `EQ_VALIDATOR_URL` with the value of `http://localhost:5001/validate`
    - **This points publisher to the local `eq-q-v`**
    - **NOTE: After changing .env variables make sure you rerun `yarn develop`**

- Import postman collection from Trello ticket into Postman
- Validate both schemas by hitting SEND individually in Postman
    - should be using a URL such as `http://localhost:9000/publish/validate`
- If errors are returned, leave a comment with the error code
- If no errors returned, validation has passed 
- Add converted questionnaire from Postman to `eq-q-r` under **/test_schemas**
- After saving, run:
    - `docker-compose build` 
    - `docker-compose up -d` or `docker-compose up`
- Load `localhost:8000` and select test file
- All Metadata should be piped through from 1 - 32. Through intro content and section intro.
- All Answers should pipe the entered date values into the title of the next question
- If any values missing or `eq-q-r` errors appear, leave a comment.
- If it's all good, you made it! 
- ... and so did the schemas

## Known issues or to be discussed
- Unable to add alternative currency types as of yet
- Date range piping bug here - https://trello.com/c/NilBm0MY/903-piping-date-range-answers
- Metadata CalcSum bug
- Multiple transformations yet

## Useful links
- https://github.com/ONSdigital/eq-schema-validator/blob/v3/doc/decisions/0001-define-piping-in-a-structured-way.md
- List of transforms 
https://github.com/ONSdigital/eq-questionnaire-validator/tree/master/schemas/string_interpolation/transforms
